hadoop-ozone/ozone-recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/FileSizeCountTask.java

```diff
@@ -0,0 +1,254 @@
+
+  @Inject
+  public FileSizeCountTask(OMMetadataManager omMetadataManager,
+      Configuration sqlConfiguration) {
+    super("FileSizeCountTask");
+    try {
+      tables.add(omMetadataManager.getKeyTable().getName());
+      fileCountBySizeDao = new FileCountBySizeDao(sqlConfiguration);
+    } catch (Exception e) {
+      LOG.error("Unable to fetch Key Table updates ", e);
+    }
+  }
+
+  protected long getOneKB() {
+    return ONE_KB;
+  }
+
+  protected long getMaxFileSizeUpperBound() {
+    return maxFileSizeUpperBound;
+  }
+
+  protected int getMaxBinSize() {
+    return maxBinSize;
+  }
+
+  /**
+   * Read the Keys from OM snapshot DB and calculate the upper bound of
+   * File Size it belongs to.
+   *
+   * @param omMetadataManager OM Metadata instance.
+   * @return Pair
+   */
+  @Override
+  public Pair<String, Boolean> reprocess(OMMetadataManager omMetadataManager) {
+    LOG.info("Starting a 'reprocess' run of FileSizeCountTask.");
+
+    fetchUpperBoundCount("reprocess");
+
+    Table<String, OmKeyInfo> omKeyInfoTable = omMetadataManager.getKeyTable();
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+        keyIter = omKeyInfoTable.iterator()) {
+      while (keyIter.hasNext()) {
+        Table.KeyValue<String, OmKeyInfo> kv = keyIter.next();
+        countFileSize(kv.getValue());
+      }
+    } catch (IOException ioEx) {
+      LOG.error("Unable to populate File Size Count in Recon DB. ", ioEx);
+      return new ImmutablePair<>(getTaskName(), false);
+    } finally {
+      populateFileCountBySizeDB();
+    }
+
+    LOG.info("Completed a 'reprocess' run of FileSizeCountTask.");
+    return new ImmutablePair<>(getTaskName(), true);
+  }
+
+  void setMaxBinSize() {
+    maxBinSize = (int)(long) (Math.log(getMaxFileSizeUpperBound())
+        /Math.log(2)) - 10;
+    maxBinSize += 2;  // extra bin to add files > 1PB.
+  }
+
+  void fetchUpperBoundCount(String type) {
+    setMaxBinSize();
+    if (type.equals("process")) {
+      //update array with file size count from DB
+      List<FileCountBySize> resultSet = fileCountBySizeDao.findAll();
+      int index = 0;
+      if (resultSet != null) {
+        for (FileCountBySize row : resultSet) {
+          upperBoundCount[index] = row.getCount();
+          index++;
+        }
+      }
+    } else {
+      upperBoundCount = new long[getMaxBinSize()];    //initialize array
+    }
+  }
+
+  @Override
+  protected Collection<String> getTaskTables() {
+    return tables;
+  }
+
+  /**
+   * Read the Keys from update events and update the count of files
+   * pertaining to a certain upper bound.
+   *
+   * @param events Update events - PUT/DELETE.
+   * @return Pair
+   */
+  @Override
+  Pair<String, Boolean> process(OMUpdateEventBatch events) {
+    LOG.info("Starting a 'process' run of FileSizeCountTask.");
+    Iterator<OMDBUpdateEvent> eventIterator = events.getIterator();
+
+    fetchUpperBoundCount("process");
```

Comment:
This code is very hard to follow, it can be simplified a bit.
1. Since Math.log2 is hardcoded and not something configurable, meaning take log to base 10 or base 2, the total number of bins is fixed anyways, right?
2. Process and reprocess can initialize the array instead of fetchUpperBoundCount, IMO

Reply:
-Yes, total number of bins is 42, based on max file size permitted (1 PB).
-For reprocess, fetchUpperBoundCount() initializes it based on maxFileSize.
For process, fetchUpperBoundCount() first initializes the array to fetch count from DB and then updates based on events for process().

Reply:
Yes, but for readability, fetchUpperBoundCount does not make sense. Also, Task should be stateless so why are we doing a setMaxBinSize?

Reply:
Sure, removed setMaxBinSize() and refactored fetchUpperBoundCount() to be only used for process() for updating count.